File: codes/array/59.spiral-matrix-ii/spiral-matrix-ii.cpp
#include <vector>
using namespace std;
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> ans(n, vector<int>(n));
        int startx = 0, starty = 0; // start position of each loop
        int loop = n / 2; // 控制循环次数
        int offset = 1; // border control
        int count = 1;

        // 注意数组ans[y][x]，即第一个是控制列的，第二个才是控制行的
        while (loop--) {
            int i, j;
            for (i = startx; i < n - offset; i++) {
                ans[starty][i] = count++;
            }
            for (j = starty; j < n - offset; j++) {
                ans[j][i] = count++;
            }
            for (; i > startx;i--) {
                ans[j][i] = count++;
            }
            for (; j > starty; j--) {
                ans[j][i] = count++;
            }
            offset++;
            startx += 1;
            starty += 1;
        }
        // 奇数情况需要补上中间的元素
        if(n%2!=0){
            ans[starty][startx] = count;
        }

        return ans;
    }
};

File: codes/array/59.spiral-matrix-ii/spiral-matrix-ii.go
package main

func generateMatrix(n int) [][]int {
	// go的多维slice的初始化只能这样，内层的需要遍历初始化
	ans := make([][]int, n)
	for i := range ans {
		ans[i] = make([]int, n)
	}

	startX := 0
	startY := 0
	loop := n / 2 // int的除法仍为int
	count := 1
	offset := 1

	for ; loop > 0; loop-- {
		i := startX
		j := startY
		for ; i < n-offset; i++ {
			ans[j][i] = count
			count++
		}

		for ; j < n-offset; j++ {
			ans[j][i] = count
			count++
		}

		for ; i > startX; i-- {
			ans[j][i] = count
			count++
		}

		for ; j > startY; j-- {
			ans[j][i] = count
			count++
		}

		startX++
		startY++
		offset++

	}
	if n%2 == 1 {
		ans[startY][startX] = count
	}

	return ans
}
